### PR TITLE
Don't use tasks to track completion of the pipe

### PR DESCRIPTION
**Pull request by davidfowl**
/cc @pakrym @halter73 @benaadams

### CODE REVIEW

**Comment by benaadams** [comment]
Made sense before; but this makes even more sense. 

Ignoring the switching of reader and writer... 😉

**Comment by pakrym** [comment]
I like it. But we can decrease size of pipe object by storing only exception and having marker exception for completed state.

**Comment by benaadams** [comment]
`NoException` 😉

**Comment by davidfowl** [comment]
I considered that but I didn't want to go that route yet because the size of pipe isn't really a problem as we aren't making one per request.

File: src/System.IO.Pipelines/Pipe.cs [hunk 1]
```diff
@@ -645,23 +625,8 @@ private static void Reset(ref Action awaitableState)
...
-            else
-            {
-                taskCompletionSource.TrySetResult(null);
-            }
-        }
-
         private void Dispose()
         {
-            Debug.Assert(Writing.IsCompleted, "Not completed writing");
-            Debug.Assert(Reading.IsCompleted, "Not completed reading");
```
Comment (halter73) [inline src/System.IO.Pipelines/Pipe.cs]: Why did you remove these asserts?

File: src/System.IO.Pipelines/Pipe.cs [hunk 2]
```diff
@@ -713,18 +678,41 @@ bool IWritableBufferAwaiter.GetResult()
...
+
+            public void TryComplete(Exception exception = null)
+            {
+                // Set the exception object to the exception passed in or a sentinel value
+                Interlocked.CompareExchange(ref _exception, exception ?? _completedNoException, null);
+            }
+
+            public void ThrowIfFailed()
+            {
+                if (_exception != null && _exception != _completedNoException)
```
Comment (halter73) [inline src/System.IO.Pipelines/Pipe.cs]: Why not just initialize _exception to be _completedNoException?

Comment (halter73) [inline src/System.IO.Pipelines/Pipe.cs]: Nvm. CompareExchange duh.

File: src/System.IO.Pipelines/Pipe.cs (hunk 1 above)
Comment (davidfowl) [inline src/System.IO.Pipelines/Pipe.cs]: laziness

Comment (halter73) [inline src/System.IO.Pipelines/Pipe.cs]: Were tests failing?
